app/src/main/java/io/apicurio/registry/storage/importing/v2/SqlDataUpgrader.java

```diff
@@ -0,0 +1,364 @@
+                    .configuration(entity.configuration).groupId(entity.groupId).build();
+            storage.importArtifactRule(newEntity);
+            log.debug("Artifact rule imported successfully: {}", entity);
+        } catch (Exception ex) {
+            log.warn("Failed to import artifact rule {}: {}", entity, ex.getMessage());
+        }
+    }
+
+    @Override
+    public void importArtifactVersion(ArtifactVersionEntity entity) {
+        try {
+            // Content needs to be imported before artifact version
+            if (!contentIdMapping.containsKey(entity.contentId)) {
+                // Add to the queue waiting for content imported
+                waitingForContent.add(entity);
+                return;
+            }
+
+            entity.contentId = contentIdMapping.get(entity.contentId);
+
+            var oldGlobalId = entity.globalId;
+            if (!preserveGlobalId) {
+                entity.globalId = storage.nextGlobalId();
+            }
+
+            Map<String, String> artifactVersionLabels = new HashMap<>();
+
+            if (entity.labels != null) {
+                artifactVersionLabels
+                        .putAll(entity.labels.stream().collect(Collectors.toMap(label -> label, s -> null)));
+            }
+
+            if (entity.properties != null) {
+                artifactVersionLabels.putAll(entity.properties);
+            }
+
+            io.apicurio.registry.utils.impexp.v3.ArtifactVersionEntity newEntity = io.apicurio.registry.utils.impexp.v3.ArtifactVersionEntity
+                    .builder().createdOn(entity.createdOn).description(entity.description)
+                    .labels(entity.labels != null
+                        ? entity.labels.stream().collect(Collectors.toMap(label -> label, label -> null))
+                        : Collections.emptyMap())
+                    .name(entity.name).owner(entity.createdBy)
+                    .state(VersionState.fromValue(entity.state.value())).artifactId(entity.artifactId)
+                    .versionOrder(entity.versionId).modifiedOn(entity.createdOn).modifiedBy(entity.createdBy)
+                    .version(entity.version).globalId(entity.globalId).contentId(entity.contentId)
+                    .groupId(entity.groupId).build();
+
+            // If the version being imported is the first one, we have to create the artifact first
+            if (entity.versionId == 1) {
+                ArtifactEntity artifactEntity = ArtifactEntity.builder().artifactId(entity.artifactId)
+                        .artifactType(entity.artifactType).createdOn(entity.createdOn)
+                        .description(entity.description).groupId(entity.groupId).labels(artifactVersionLabels)
+                        .modifiedBy(entity.createdBy).modifiedOn(entity.createdOn).name(entity.name)
+                        .owner(entity.createdBy).build();
+                storage.importArtifact(artifactEntity);
+            }
```

Comment:
This logic will only work if the sequence of versions is in the correct order.  I'm not sure we can guarantee this once we switch from the current ZIP streaming approach to one where we unpack the ZIP and then read the entities from disk.

We may need to check for the existence of the artifact instead.  But maybe leave this for now and react later once we make other changes.